iree/hal/cuda/cuda_device.c

```diff
@@ -229,19 +266,17 @@ static iree_status_t iree_hal_cuda_device_queue_submit(
     iree_hal_queue_affinity_t queue_affinity, iree_host_size_t batch_count,
     const iree_hal_submission_batch_t* batches) {
   iree_hal_cuda_device_t* device = iree_hal_cuda_device_cast(base_device);
+  iree_hal_command_buffer_t* target_command_buffer = batches[0].command_buffers[0];
```

Comment:
This should not be pulling a command buffer from here for the target, but instead applying it against a stack-allocated or device-owned command buffer. You also don't know whether there's any batches or even command buffers within any batches (it's legal to not pass any command buffers if the submission is used only for semaphore operations).

Something like:
```
iree_hal_cuda_stream_command_buffer_t stream_command_buffer;
iree_hal_cuda_stream_command_buffer_initialize(/*cuda stream selection*/, &stream_command_buffer);
for ( ... ) {
  for ( ... ) {
    iree_hal_deferred_command_buffer_apply(batches.command_buffers..., &stream_command_buffer);
  }
}
```
or
```
// in device creation:
iree_hal_cuda_stream_command_buffer_initialize(/*cuda stream selection*/, &device->stream_command_buffers[0]);
iree_hal_cuda_stream_command_buffer_initialize(/*cuda stream selection*/, &device->stream_command_buffers[1]);
...
// in submit:
iree_hal_cuda_stream_command_buffer_t* stream_command_buffer = device->stream_command_buffers[/*<queue_affinity>*/];
for ( ... ) {
  for ( ... ) {
    iree_hal_deferred_command_buffer_apply(batches.command_buffers..., stream_command_buffer);
  }
}
```

If the command buffer is mostly stateless (beyond recording state) then allocating on the stack is fine size-wise, but synchronization will be important regardless: submissions are atomic and you must synchronize access to the stream so that if another thread performs a submit at the same time they don't stop on each other's use of the same underlying cuda stream. You could do this by having one iree_slim_mutex_t per stream and then after selecting the stream based on the queue_affinity hold the lock while applying the deferred commands.